docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,310 @@
+* `"type": "type name"` the type of the global value
+* optional `"value": VALUE | "unknown"` the value of the global value or "unknown". If omitted, same as "unknown".
+
+Note that the logical descriptor does not contain "unknown" values.
+
+The `VALUE` may be a JSON numeric constant integer or a string containing a signed or unsigned
+decimal or hex (with prefix `0x` or `0X`) integer constant.  The constant must be within the range
+of the type of the global value.
+
+For pointer and nuint globals, the value may be assumed to fit in a 64-bit unsigned integer.  For
+nint globals, the value may be assumed to fit in a 64-bit signed integer.
+
+If the value is specified as "unknown" another physical descriptor in the "baseline" graph is
+expected to provide the value.
+
+## Physical binary blob descriptor
+
+### Version
+
+This is version 0 of the physical binary blob format.
+
+### Summary
+
+The binary blob format for a physical descriptor is expected to be stored in the memory space of a
+target .NET runtime and is thus likely to be the final descriptor in the "baseline" graph.
+
+It is likely that the physical descriptor will be a part of a build-time constant in the disk image
+of a .NET runtime, and the format is designed to be compact and specifiable as a (likely
+machine-generated) compile-time constant in a suitable source language.
+
+The data descriptor forms one part of an overall physical data contract descriptor in a targret .NET
+runtime and as such this format does not specify a "magic number", or "well known symbol", or
+another means of identifying the blob within a target process.  Additionally the version of the
+binary blob data descriptor is expected to be stored within the larger enclosing data contract
+descriptor and is not included here.
+
+### Blob
+
+Multi-byte values are in the target platform endianness.
+
+The format is:
+
+```c
+struct BinaryBlobDataDescriptor
+{
+    struct Directory {
+        uint32_t TypesStart;
+        uint32_t FieldPoolStart;
+        
+        uint32_t GlobalValuesStart;
+        uint32_t NamesStart;
+        
+        uint32_t TypeCount;
+        uint32_t FieldPoolCount;
+        
+        uint32_t NamesPoolCount;
+        uint32_t Reserved0;
+        
+        uint16_t TypeSpecSize;
+        uint16_t FieldSpecSize;
+        
+        uint16_t GlobalSpecSize;
+        uint16_t Reserved1;
+    }
+    uint32_t BaselineName;
+    TypeSpec[TypeCount] Types;
+    FieldSpec[FieldPoolCount] FieldPool;
+    GlobalSpec[GlobalsCount] GlobalValues;
+    uint8_t[NamesPoolCount] NamesPool;
+};
+
+struct TypeSpec
+{
+    uint32_t Name;
+    uint32_t Fields;
+    uint16_t Size;
+};
+
+struct FieldSpec
+{
+    uint32_t Name;
+    uint16_t FieldOffset;
+    uint16_t Reserved;
+};
+
+struct GlobalSpec
+{
+    uint32_t Name;
+    uint32_t Reserved;
+    uint64_t Value;
+};
+```
+
+Rationale: the blob should be producable by including a specially formatted C header file multiple
```

Comment:
> the blob should be producable by including a specially formatted C header file multiple

What's the rationale for this requirement?

Reply:
> What's the rationale for this requirement?

This comes from two requirements:

1. We want to be able to read off a binary descriptor from an object file by just looking at the object file as a sequence of bytes - we don't want to know anything about symbols, relocations, dwarf info, etc. So the binary descriptor essentially has to be a `const uint8_t blob[] = {...}`
2. For the most restricted runtime (Mono) we must be able to produce the blob using only what is available in a C compiler.

Together, these imply that we want to produce a self-describing blob (ie: if the C compiler decides to add padding or alignment, we must be able to ignore them) using only the C preprocessor.

That doesn't mean we _must_ use a C preprocessor to create the blob - a NativeAOT compiler can emit the blob using other means, and a runtime written in C++ can use templates or `constexpr` to create the data.

Reply:
FWIW, both these requirements can be fulfilled by the json format too. 

The only problem are absolute pointers. They would have to be expressed as indices into auxiliary array. The same problem exists for `const uint8_t blob[] = {...}` too, but what you are proposing is not actually `const uint8_t blob[] = ...`.